### PR TITLE
add a AbstractMetadataClient and make MetadataClient to extends it

### PR DESCRIPTION
**Pull request by yug095**
Make this change to CDAP for apps to create a discovery metadata client to extends the abstract one. 

http://builds.cask.co/browse/CDAP-DUT4461-1

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java [hunk 1]
```diff
@@ -0,0 +1,1159 @@
...
+  }
+
+  private String constructPath(Id.Stream streamId) {
+    return String.format("streams/%s", streamId.getId());
+  }
+
+  private String constructPath(Id.Stream.View viewId) {
+    return String.format("streams/%s/views/%s", viewId.getStreamId(), viewId.getId());
+  }
+}
```
Comment (anwar6953) [inline cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java]: have newline at end of file

File: cdap-client/src/main/java/co/cask/cdap/client/MetadataClient.java [hunk 2]
```diff
@@ -70,1093 +46,18 @@ public MetadataClient(ClientConfig config) {
...
   }
 
-  private String constructPath(Id.Stream.View viewId) {
-    return String.format("streams/%s/views/%s", viewId.getStreamId(), viewId.getId());
+  @Override
+  protected URL resolve(Id.Namespace namespace, String resource) throws MalformedURLException {
+    return config.resolveNamespacedURLV3(namespace, resource);
   }
-}
+}
```
Comment (anwar6953) [inline cdap-client/src/main/java/co/cask/cdap/client/MetadataClient.java]: same here

File: cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java [hunk 3]
```diff
@@ -0,0 +1,1159 @@
...
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Common implementation of methods to interact with metadata over HTTP.
+ */
+public abstract  class AbstractMetadataClient {
```
Comment (anwar6953) [inline cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java]: nit - extra space after abstract

File: cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java [hunk 4]
```diff
@@ -0,0 +1,1159 @@
...
+  private static final Type SET_METADATA_RECORD_TYPE = new TypeToken<Set<MetadataRecord>>() { }.getType();
+  private static final Type MAP_STRING_STRING_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+  private static final Type SET_STRING_TYPE = new TypeToken<Set<String>>() { }.getType();
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Id.NamespacedId.class, new NamespacedIdCodec()).create();
+
+  /**
+   * Executes an HTTP request.
+   */
+  protected abstract  HttpResponse execute(HttpRequest request,  int... allowedErrorCodes)
```
Comment (anwar6953) [inline cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java]: nit - extra space after `abstract` and before `int`

**Comment by anwar6953** [comment]
Changed overall seem good, in order to support other implementations of metadata client (i.e. via discovery).
@bdmogal ?

File: cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java [hunk 5]
```diff
@@ -0,0 +1,1159 @@
...
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Common implementation of methods to interact with metadata over HTTP.
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java]: ... interact with metadata service over HTTP

**Comment by bdmogal** [comment]
This is a little hard to review, wish there was a better way for github to show this diff. However, as long as this is a simple cut-paste from `MetadataClient` to `AbstractMetadataClient`, with the 2 additional abstract methods, and tests pass, LGTM

File: cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java (hunk 5 above)
Comment (yug095) [inline cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java]: added

File: cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java (hunk 4 above)
Comment (yug095) [inline cdap-common/src/main/java/co/cask/cdap/common/metadata/AbstractMetadataClient.java]: changed

**Pull request merged**